tensorflow_privacy/privacy/membership_inference_attack/codelabs/privacy_risk_score_codelab.ipynb

```diff
@@ -0,0 +1,804 @@
+      "text/plain": [
+       "<Figure size 432x288 with 1 Axes>"
+      ]
+     },
+     "metadata": {
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "from tensorflow_privacy.privacy.membership_inference_attack.data_structures import AttackInputData\n",
+    "from tensorflow_privacy.privacy.membership_inference_attack.data_structures import SlicingSpec\n",
+    "from tensorflow_privacy.privacy.membership_inference_attack.data_structures import AttackType\n",
+    "\n",
+    "import tensorflow_privacy.privacy.membership_inference_attack.plotting as plotting\n",
+    "\n",
+    "labels_train = np.argmax(y_train, axis=1)\n",
+    "labels_test = np.argmax(y_test, axis=1)\n",
+    "\n",
+    "input = AttackInputData(\n",
+    "  logits_train = logits_train,\n",
+    "  logits_test = logits_test,\n",
+    "  loss_train = loss_train,\n",
+    "  loss_test = loss_test,\n",
+    "  labels_train = labels_train,\n",
+    "  labels_test = labels_test\n",
+    ")\n",
+    "\n",
+    "# Run several attacks for different data slices\n",
+    "attacks_result = mia.run_attacks(input,\n",
+    "                                 SlicingSpec(\n",
+    "                                     entire_dataset = True,\n",
+    "                                     by_class = True,\n",
+    "                                     by_classification_correctness = True\n",
+    "                                 ),\n",
+    "                                 attack_types = [\n",
+    "                                     AttackType.THRESHOLD_ATTACK,\n",
+    "                                     AttackType.THRESHOLD_ENTROPY_ATTACK,\n",
+    "                                     AttackType.LOGISTIC_REGRESSION])\n",
+    "\n",
+    "# Plot the ROC curve of the best classifier\n",
+    "fig = plotting.plot_roc_curve(\n",
+    "    attacks_result.get_result_with_max_auc().roc_curve)\n",
+    "\n",
+    "# Print a user-friendly summary of the attacks\n",
+    "print(attacks_result.summary(by_slices = True))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "E9zwsPGFujVq"
+   },
+   "source": [
+    "## Compute privacy risk score\n",
+    "\n",
+    "This part shows how to use the privacy risk score.\n",
+    "\n",
+    "For each data slice, we compute privacy risk scores for both training and test data. We then set a threshold on risk scores (an input is inferred as a member if and only if its risk score is higher than the threshold) and compute the attack precision and recall values"
```

Comment:
Would that be possible to add some interpretation of the obtained results?

Some questions that we could highlight:
 - How does the precision / recall figures compare to the membership inference attack? As you have the results for MIA right above, it might be helpful to compare the two methods. 
- Are there any samples that have high risk scores? Is there anything special about those samples?
 - What is the distribution of privacy risk scores? We could probably plot a simple histogram.

Reply:
For the precision-recall figures, it will be pretty similar to the threshold attacks based on prediction loss or entropy since the privacy risk score is computed based on the distributions of prediction loss or entropy over training and test data.
The importance of privacy risk score analysis is that we actually compute a risk value for each sample such that we can know which samples have high risks. The precision-recall metric is just one way to present the results.

Indeed, some samples have high risk scores. In the codelab running example, class 3 and class 4 have certain training samples with the risk score of 1. It is a very interesting future direction to explore why those samples have high risks.

I agree that we can plot the histogram of the privacy risk scores.
I will suggest a meeting after the holiday season to thoroughly discuss what is the best way to present the privacy risk score results.

Reply:
Setting up a meeting sounds good, thank you. Some comments (not blocking for this PR, we can address those after the meeting):

`it will be pretty similar to the threshold attacks based on prediction loss or entropy since the privacy risk score is computed based on the distributions of prediction loss or entropy over training and test data.`

This sounds reasonable. Can we compare these numbers in the codelab, as a sanity check?

`It is a very interesting future direction to explore why those samples have high risks.`

As a low hanging fruit, could we plot a few pictures with low and high risk? That might already be very insightful.
We're looking for some way to show how this metrics work in practice.

`importance of privacy risk score analysis is that we actually compute a risk value for each sample such that we can know which samples have high risks`

This makes me realize that some of the trained attacks, and the threshold attack also give some kind of score, which we threshold to classify into training / non-training examples. For the threshold attack, this score is simply the loss. For the neural network classifiers, this is the output of the softmax layer.

Conceptually your score a very similar, except that you provide a different empirical measure.

We discussed this internally, and we think it might make sense to consolidate this with the other membership attacks instead of having this as a separate codepath. This might make the implementation cleaner and simpler.

We're happy to do this refactoring ourselves, just wanted to let you know about this.

Reply:
I added the AUC and advantage values into the new codelab file. As expected, the results are similar to threshold attacks.

I also plotted several figures with high and low risks for each class label in the codelab file, please take a look.

I agree that neural network attack classifiers or threshold attacks can also give some kind of scores.
The advantage of our proposed privacy risk score metric is that it closely captures the real likelihood of being in the training set. For example, if we count all training and test samples with the privacy risk score around (let's say) 0.8, then among those samples, it is indeed around 80% of them are from the training set. While neural network classifier output usually does not well capture the real likelihood of being a member. You can check more in our USENIX paper (Figure 3 and Figure 11 in https://arxiv.org/pdf/2003.10595.pdf)

Sure, I can see the benefits of consolidating the privacy risk score part with other attacks.